File: PD Project/Firmware/PD_VRMT_v1_0_6_20240814/main/HX711.cpp
#include <Arduino.h>
#include "HX711.h"

HX711::HX711() {};

void HX711::init(int pin_dout, int pin_pd_sck) {
    _pin_dout = pin_dout;
    _pin_pd_sck = pin_pd_sck;
    pinMode(_pin_dout, INPUT);
    pinMode(_pin_pd_sck, OUTPUT);
}

unsigned long HX711::read() {
    unsigned long count = 0; 

    // when output data is not ready for retrieval
    // digital output pin DOUT is high
    // serial clock input should be low
	digitalWrite(_pin_dout, HIGH);
	delayMicroseconds(1);
	digitalWrite(_pin_pd_sck, LOW);
	delayMicroseconds(1);

  	
    // when DOUT goes to low, it indictes data is ready for retrieval
	while(digitalRead(_pin_dout));

    // by applying 25-27 positive clock pulses at the PD_SCK pin
    // data is shifted out from the DOUT output pin
    // each pulse shifts out one bit, starting at MSB bit first
    // until all 24 bits are shifted out
	for(int i = 0; i < 24; i++){ 
		// pulse high
        digitalWrite(_pin_pd_sck, HIGH); 
		delayMicroseconds(1);
		
        // shift the bit too the left
        count = count << 1; 
		
        // pulse low
        digitalWrite(_pin_pd_sck, LOW); 
		delayMicroseconds(1);

        // data out, registered at the right-most bit
		if(digitalRead(_pin_dout))
			count++; 
	} 

    // the 25th bit will pull DOUT pin back to high
	digitalWrite(_pin_pd_sck, HIGH); 
	delayMicroseconds(1);
	digitalWrite(_pin_pd_sck, LOW); 
	delayMicroseconds(1);
	
    // the output 24 bits of data is in 2's complement format
    // when goes out of range, output data will be saturated at 800000h (min) or 7fffffh (max)
	count ^= 0x800000;
	return count;
}

unsigned long HX711::set_zero() {
    // get ADC reading at zero weight
    _zero = read();
    return _zero;
}

unsigned long HX711::get_zero() {
    return _zero;
}

unsigned long HX711::set_cal() {
    // get ADC reading at 1kg weight
    _cal = (read() - _zero) / 1000.0;
    return _cal;
}

unsigned long HX711::get_cal() {
    return _cal;
}

void HX711::input_zero(float val) {
    _zero = val;
}

void HX711::input_cal(float val) {
    _cal = val;
}

float HX711::get_weight() {
    // compute calibrated weighted load cell
    weight = (float) ((read() - _zero) / _cal);
    return weight;
}

File: PD Project/python_code/hx711.py
import RPi.GPIO as GPIO
import time

class HX711:
    def __init__(self, pin_dout, pin_pd_sck):
        self._pin_dout = pin_dout
        self._pin_pd_sck = pin_pd_sck
        self._zero = 9000000.0   # Default tare (zero) value
        self._cal = 870.0        # Default calibration factor
        self.weight = 0.0 
        
        GPIO.setmode(GPIO.BCM)
        GPIO.setup(self._pin_dout, GPIO.IN)
        GPIO.setup(self._pin_pd_sck, GPIO.OUT)
        GPIO.output(self._pin_pd_sck, GPIO.LOW)

    # Read data
    def read(self):
        count = 0 #initialize count to store 24-bits reading (all bits 0)
        
        while GPIO.input(self._pin_dout) == GPIO.HIGH: # Loop exit when DOUT low
            time.sleep(0.001)
        
        for i in range(24):#clock out 24 bits
            GPIO.output(self._pin_pd_sck, GPIO.HIGH) # when clock pulse high, read output
            time.sleep(0.000001)
            count = count << 1 #left-shift operation (new bit shifted into LSB)
            GPIO.output(self._pin_pd_sck, GPIO.LOW)
            time.sleep(0.000001)
            
            if GPIO.input(self._pin_dout) == GPIO.HIGH:
                count += 1 
        
        # After read 24 bits
        GPIO.output(self._pin_pd_sck, GPIO.HIGH)# 25th pulse starts 
        time.sleep(0.000001)
        GPIO.output(self._pin_pd_sck, GPIO.LOW) # 25th pulse starts 
        time.sleep(0.000001)
        
        count ^= 0x800000 # turns from 2's complement to offset binary(only positive value)
        return count

    # Calibration
    def set_zero(self):
        self._zero = self.read()
        return self._zero

    def get_zero(self):
            return self._zero

    def set_cal(self, known_weight=1000.0):
        reading = self.read()
        self._cal = (reading - self._zero) / known_weight
        return self._cal

    def get_cal(self):
            return self._cal 

    def input_zero(self, val):
            self._zero = val

    def input_cal(self, val):
            self._cal = val


    # Weight calculation
    def get_weight(self):
        self._weight = (self.read() - self._zero) / self._cal
        return self._weight


    def cleanup(self):
        GPIO.cleanup()
